internal/telemetry/reporter.go

```diff
@@ -0,0 +1,232 @@
+	}
+
+	r.Header.Add("X-Prometheus-Remote-Write-Version", "0.1.0")
+	r.Header.Add("Content-Encoding", "snappy")
+	r.Header.Set("Content-Type", "application/x-protobuf")
+
+	resp, err := client.Do(r)
+	if err != nil {
+		return fmt.Errorf("failed to send Prometheus remote write: %w", err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode/100 != 2 {
+		body, _ := ioutil.ReadAll(resp.Body)
+		return fmt.Errorf(
+			"unexpected Prometheus remote write response: %d: %s",
+			resp.StatusCode,
+			string(body),
+		)
+	}
+
+	return nil
+}
+
+func discoverTimeseries(registry *prometheus.Registry) (allTS []*prompb.TimeSeries, err error) {
+	metricFams, err := registry.Gather()
+	if err != nil {
+		return nil, fmt.Errorf("failed to gather telemetry metrics: %w", err)
+	}
+
+	defaultTimestamp := model.Time(time.Now().UnixNano() / int64(time.Millisecond))
+	sampleVector, err := expfmt.ExtractSamples(&expfmt.DecodeOptions{
+		Timestamp: defaultTimestamp,
+	}, metricFams...)
+	if err != nil {
+		return nil, fmt.Errorf("unable to extract sample from metrics families: %w", err)
+	}
+
+	for _, sample := range sampleVector {
+		allTS = append(allTS, &prompb.TimeSeries{
+			Labels: convertLabels(sample.Metric),
+			Samples: []*prompb.Sample{{
+				Value:     float64(sample.Value),
+				Timestamp: int64(sample.Timestamp),
+			}},
+		})
+	}
+
+	return
+}
+
+func discoverAndWriteMetrics(
+	ctx context.Context,
+	registry *prometheus.Registry,
+	client *http.Client,
+	endpoint string,
+) error {
+	ts, err := discoverTimeseries(registry)
+	if err != nil {
+		return err
+	}
+
+	return writeTimeSeries(ctx, client, endpoint, ts)
+}
+
+type Reporter func(ctx context.Context) error
+
+// RemoteReporter creates a telemetry reporter with the specified parameters, or errors
+// if the configuration was invalid.
+func RemoteReporter(
+	registry *prometheus.Registry,
+	endpoint string,
+	caOverridePath string,
+	interval time.Duration,
+) (Reporter, error) {
+	if _, err := url.Parse(endpoint); err != nil {
+		return nil, fmt.Errorf("invalid telemetry endpoint: %w", err)
+	}
+	if interval < MinimumAllowedInterval {
+		return nil, fmt.Errorf("invalid telemetry reporting interval: %s < %s", interval, MinimumAllowedInterval)
+	}
+	if endpoint == DefaultEndpoint && interval != DefaultInterval {
+		return nil, fmt.Errorf("cannot change the telemetry reporting interval for the default endpoint")
+	}
+
+	client := &http.Client{}
+	if caOverridePath != "" {
+		pool, err := x509util.CustomCertPool(caOverridePath)
+		if err != nil {
+			return nil, fmt.Errorf("invalid custom cert pool path `%s`: %w", caOverridePath, err)
+		}
+
+		t := &http.Transport{
+			TLSClientConfig: &tls.Config{
+				RootCAs:    pool,
+				MinVersion: tls.VersionTLS12,
+			},
+		}
+
+		client.Transport = t
+	}
+
+	return func(ctx context.Context) error {
+		// Smear the startup delay out over 10% of the reporting interval
+		startupDelay := time.Duration(rand.Int63n(int64(interval.Seconds()/10))) * time.Second
```

Comment:
Since that could be up to 10m after startup, maybe we should attempt to write on shutdown if we haven't written at all yet and the server has at least successfully started up. And actually, maybe we want separate telemetry on startup error rates. 

I realize this is a fairly involved change for an uncommon edge case, maybe we just make an issue and do it later.